dagshub/streaming/filesystem.py

```diff
@@ -482,38 +469,95 @@ def scandir(self, path='.'):
+                                    is_directory=is_directory, is_binary=is_binary)
+
+        has_storages = len(self._storages) > 0
+        res = set()
+        str_path = dh_path.relative_path.as_posix()
+        if str_path == ".":
+            res.add(generate_entry(SPECIAL_FILE, False))
+            if has_storages:
+                res.add(generate_entry(".dagshub", True))
+        elif str_path.startswith(".dagshub") and has_storages:
+            storage_paths = [s.path_in_mount for s in self._storages]
+            for sp in storage_paths:
+                try:
+                    relpath = sp.relative_to(dh_path.relative_path)
+                    if relpath != Path():
+                        res.add(generate_entry(relpath.parts[0], True))
+                except ValueError:
+                    continue
+        return res
+
+    def _api_listdir(self, path: DagshubPath, include_size: bool = False) -> Optional[List[ContentAPIEntry]]:
+        response, hit = self._check_listdir_cache(path.relative_path.as_posix(), include_size)
+        if hit:
+            return response
+        response = self.http_get(path.content_url,
+                                 params={'include_size': 'true'} if include_size else {},
+                                 headers=config.requests_headers)
+        if response.status_code >= 400:
+            logger.debug(f"Got HTTP code {response.status_code} while listing {path}, no results will be returned")
+            return None
+        res = []
+        for entry_raw in response.json():
+            entry = dacite.from_dict(ContentAPIEntry, entry_raw)
+            # Ignore storage root entries, we handle them separately in a different place
+            if entry.type == "storage":
+                continue
+            res.append(entry)
+        self._listdir_cache[(path.relative_path.as_posix(), include_size)] = res
+        return res
+
+    def _api_storages(self) -> List[StorageAPIEntry]:
+        response = self.http_get(self.storage_api_url)
+        if response.status_code >= 400:
+            logger.warning(f"Got HTTP code {response.status_code} while getting storages. Content: {response.content}")
+            logger.warning("Storages are unavailable")
+            return []
+        return [dacite.from_dict(StorageAPIEntry, a) for a in response.json()]
+
+    def _check_listdir_cache(self, path: str, include_size: bool) -> Tuple[Optional[List[ContentAPIEntry]], bool]:
+        # If we already have a response with side included, return that
+        if (path, True) in self._listdir_cache and not include_size:
+            self._listdir_cache[(path, False)] = self._listdir_cache[(path, True)]
+        if (path, include_size) in self._listdir_cache:
+            return self._listdir_cache[(path, include_size)], True
+        return None, False
+
+    def _api_download_file_git(self, path: DagshubPath):
+        return self.http_get(path.raw_url, headers=config.requests_headers, timeout=None)
```

Comment:
This creates an unecessary circular dependency between `DagshubPath` and `DagsHubFilesystem`. 
What about the following pattern:
```python
self.http_get(self.raw_url(path, path.is_storage_path))
```
`raw_url` would need to move to `filesystem.py`

Reply:
I still can't get rid of the circular dependency because I need the project_root for the storage path handling.
Might add a third constructor argument just for that 🤔